drivers/wdf/wdfldr/class.c

```diff
@@ -0,0 +1,895 @@
+
+	if (result != NULL)
+	{
+		memset(result, 0, sizeof(CLASS_CLIENT_MODULE));
+		InitializeListHead(&result->ClassLinkage);
+		InitializeListHead(&result->ClientLinkage);
+	}
+
+	return result;
+}
+
+
+PCLIENT_MODULE
+NTAPI
+LibraryFindClientLocked(
+	IN PLIBRARY_MODULE LibModule,
+	IN PWDF_BIND_INFO BindInfo
+)
+{
+	PCLIENT_MODULE result;
+
+	result = (PCLIENT_MODULE)LibModule->ClientsListHead.Flink;
+
+	for (; result != (PCLIENT_MODULE)&LibModule->ClientsListHead;
+		result = (PCLIENT_MODULE)result->LibListEntry.Flink)
+	{
+
+		if (result->Info == BindInfo)
+			return result;
+	}
+
+	return NULL;
+}
+
+
+NTSTATUS
+NTAPI
+ClassLinkInClient(
+	IN PCLASS_MODULE ClassModule,
+	IN PWDF_CLASS_BIND_INFO ClassBindInfo,
+	IN PWDF_BIND_INFO BindInfo,
+	OUT PCLASS_CLIENT_MODULE ClassClientModule
+)
+{
+	PCLIENT_MODULE pClienInfo;
+	NTSTATUS status;
+
+	ClassClientModule->Class = ClassModule;
+	ClassClientModule->ClientClassBindInfo = ClassBindInfo;
+	LibraryAcquireClientLock(ClassModule->Library);
+	pClienInfo = LibraryFindClientLocked(ClassModule->Library, BindInfo);
+
+	if (pClienInfo != NULL)
+	{
+		ClassClientModule->Client = pClienInfo;
+		InsertTailList(&pClienInfo->ClassListHead, &ClassClientModule->ClassLinkage);
+	}
+
+	LibraryReleaseClientLock(ClassModule->Library);
+
+	if (pClienInfo != NULL)
+	{
+		ClassAcquireClientLock(&ClassModule->ClientsListLock);
+		InsertTailList(&ClassModule->ClientsListHead, &ClassClientModule->ClientLinkage);
+		ClassReleaseClientLock(&ClassModule->ClientsListLock);
+		status = STATUS_SUCCESS;
+	}
+	else
+	{
+		if (WdfLdrDiags)
+		{
+			DbgPrint("WdfLdr: ClassLinkInClient - ");
+			DbgPrint("ERROR: Could not locate client from Info %p, status 0x%x\n", BindInfo, STATUS_INVALID_DEVICE_STATE);
+		}
+		status = STATUS_INVALID_DEVICE_STATE;
+	}
+
+	return status;
+}
+
+
+NTSTATUS
+NTAPI
+GetClassRegistryHandle(
+	IN PWDF_CLASS_BIND_INFO ClassBindInfo,
+	IN PWDF_BIND_INFO BindInfo,
+	OUT PHANDLE KeyHandle
+)
+{
+	PWCHAR pClassName;
+	PWCHAR pClassNameBegin;
+	WCHAR currentSym;
+	PWDF_BIND_INFO pBindInfo;
+	PWDF_CLASS_BIND_INFO pClassBindInfo;
+	UNICODE_STRING classVersions;
+	OBJECT_ATTRIBUTES ObjectAttributes;
+	HANDLE rootHandle;
+	HANDLE classVersionsHandle;
+	HANDLE classVersionHandle;
+	NTSTATUS status;
+	SIZE_T numOfBytes;
+	UNICODE_STRING objectName;
+	DECLARE_UNICODE_STRING_SIZE(versionString, 22);
+	
+	pClassBindInfo = ClassBindInfo;
+	pBindInfo = BindInfo;
+	classVersions.Length = 0;
+	classVersions.Buffer = NULL;
+	RtlInitUnicodeString(&objectName, L"\\Registry\\Machine\\System\\CurrentControlSet\\Control\\Wdf\\Kmdf");
+	ObjectAttributes.ObjectName = &objectName;
+	rootHandle = NULL;
+	classVersionsHandle = NULL;
+	classVersionHandle = 0;
+
+	*KeyHandle = NULL;
+	ObjectAttributes.Length = 24;
+	ObjectAttributes.RootDirectory = 0;
+	ObjectAttributes.Attributes = OBJ_KERNEL_HANDLE | OBJ_CASE_INSENSITIVE;//576;
+	ObjectAttributes.SecurityDescriptor = 0;
+	ObjectAttributes.SecurityQualityOfService = 0;
```

Comment:
Can you use the appropriate macro InitializeObjectAttributes() and avoid magic value?